flow.py

```diff
@@ -89,25 +89,66 @@ def __init__(self, id_=None, start=None, amount=None, src_obj=None, dst_obj=None
 
         self.time_of_last_window_reduction = float('-inf')
 
+        # Smallest sequence number that receiver hasn't received
+        self.next_expected = 0
+        self.packets_received_out_of_order = []
+
+        # Start of transmit window (earliest unacknowledged sequence number)
+        self.transmit_window_start = 0
+
+        # If we eventually do selective acknowledgements
+        #self.packets_ackd_out_of_order = []
+
+        # TCP States (affect how window is changed)
+        self.slow_start = True
+        self.congestion_avoidance = False
+        self.ssthresh = float('inf')
+
     def init_window_size(self):
         self.window_size = 1
+        self.window_size_float = float(1)
 
+    # This is called when waiting to receive an ack for a sent packet times out
     def update_window_size_missed_ack(self):
         '''
-        Cut the window size in half, unless the window size was halved within
-        the previous 0.5 seconds
+        Set ssthresh to half of the current window size,
+        unless a timeout happened within the past 0.5 seconds or rtte,
+        then set the window size to 1
         '''
         time = globals_.event_manager.get_time()
-        if time - self.time_of_last_window_reduction > 0.5:
+        if time - self.time_of_last_window_reduction > max(0.5, self.rtte.estimate):
             self.time_of_last_window_reduction = time
             new_size = (self.window_size // 2) + (self.window_size % 2)
             assert new_size >= 1
-            self.set_window_size(new_size)
+            # DEBUG(jg)
+            globals_.event_manager.log('WINDOW updating ssthresh to {}'.format(new_size))
+            # ENDEBUG
+            self.ssthresh = new_size
+            self.set_window_size(1)
 
+    def retransmit(self):
+        self.packets_waiting_for_acks = dict()
+        self.act()
```

Comment:
I think `self.act()` should be removed and replaced with a comment that explains the following:

When the event manager finishes processing the `ack_is_due` event, it will call `actor.act()` for each actor. So, it is unnecessary to put `self.act()` here -- the event manager will call it. 

It is important for the event manager to get the correct return value from act() [should return true if "something happened," and false otherwise]. By calling `self.act()` here, we might cause something to happen, but the event manager won't know that anything happened. Then, the event manager will call act(), but will nothing will happen. In summary: Only the event manager should be calling act().